### PR TITLE
Landing page controller

### PR DESCRIPTION
**Pull request by rap1ds**
Base controller for landing page. It contains the Hero section, some sample data and the data denormalizer.

The controller is unaccessible at the moment, but if you want to test it, go to routes.rb and uncomment the landing page routes.

This PR doesn't contain the assets needed for testing the landing page (`public/landing_page` folder missing). Ping me if you need these.

### CODE REVIEW

File: app/utils/hash_utils.rb [hunk 1]
```diff
@@ -144,4 +144,39 @@ def flatten(h)
...
+  #
+  # Example (stringify keys):
+  #
+  # deep_map(a: 1, b: 2) { |k, v|
+  #   [k.to_s, v]
+  # }
+  #
+  # Unlike Ruby's Hash#map, this method returns a Hash, not an Array.
+  #
+  def deep_map(obj, &block)
```
Comment (ovan) [inline app/utils/hash_utils.rb]: I'm again thinking about naming and what's the general case here. This is similar to tree walking except that in a typical tree walk you can process the non-leaf nodes too if you want to. This impl would be something like `prewalk-leaves`, which really is just a special case of `prewalk`. See e.g. https://clojure.github.io/clojure/clojure.walk-api.html

Comment (ovan) [inline app/utils/hash_utils.rb]: Actually I read this wrong. This does allow you to process all nodes.

Comment (rap1ds) [inline app/utils/hash_utils.rb]: I don't know what is the actual definition of "walking", but to me it sounds more generic than this operation, which is "mapping". To me "walking" sounds like iterating all the nodes, but doesn't specify what the operation is (e.g. replacing all the `.map` calls in this function with `.each`). But again, I'm not sure what's the correct definition. Do you know?

File: app/controllers/landing_page_controller.rb [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+    end
+
+    def to_tree(normalized_data)
+      root = normalized_data[@root]
+
+      HashUtils.deep_map(root) { |k, v|
+        case v
+        when Hash
+          type = v["type"]
+          id = v["id"]
```
Comment (ovan) [inline app/controllers/landing_page_controller.rb]: `type, id = v.values_at("type", "id")`

File: app/controllers/landing_page_controller.rb [hunk 3]
```diff
@@ -0,0 +1,110 @@
+class LandingPageController < ActionController::Metal
+
+  class Denormalizer
+
+    def initialize(root: "composition", hooks: {})
+      @root = root
+      @hooks = hooks
```
Comment (ovan) [inline app/controllers/landing_page_controller.rb]: So the hooks are post-processors for linked data?

File: app/controllers/landing_page_controller.rb (hunk 2 above)
Comment (rap1ds) [inline app/controllers/landing_page_controller.rb]: TIL

File: app/controllers/landing_page_controller.rb (hunk 3 above)
Comment (rap1ds) [inline app/controllers/landing_page_controller.rb]: Yeah, that's the only thing they do at the moment. To be honest I didn't think too much about hooks, but just needed some way to add the proper path to the images.

**Pull request merged**